{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7e1a8a6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Relative Frequency\n",
      "ATG            0.027185\n",
      "CAT            0.016198\n",
      "CAA            0.030352\n",
      "ACA            0.027369\n",
      "TCT            0.015518\n",
      "..                  ...\n",
      "CCG            0.006812\n",
      "GTC            0.005476\n",
      "TAG            0.000598\n",
      "CGG            0.001357\n",
      "TGA            0.000599\n",
      "\n",
      "[64 rows x 1 columns]\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import CodonUsage\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "\n",
    "folder_path = 'C:/Users/seq'\n",
    "filename = 'cds_from_genomic_thuringiensis.fna'\n",
    "file_path = os.path.join(folder_path, filename)\n",
    "\n",
    "def load_genome(file_path):\n",
    "    sequences = []\n",
    "    with open(file_path, 'r') as file:\n",
    "        for record in SeqIO.parse(file, \"fasta\"):\n",
    "            sequences.append(str(record.seq))\n",
    "    return sequences\n",
    "\n",
    "def calculate_codon_usage(sequences):\n",
    "    codon_usage = defaultdict(int)\n",
    "    total_codons = 0\n",
    "    \n",
    "    for sequence in sequences:\n",
    "        for i in range(0, len(sequence) - 2, 3):\n",
    "            codon = sequence[i:i+3]\n",
    "            if len(codon) == 3:\n",
    "                codon_usage[codon] += 1\n",
    "                total_codons += 1\n",
    "\n",
    "    for codon in codon_usage:\n",
    "        codon_usage[codon] /= total_codons\n",
    "\n",
    "    return pd.DataFrame.from_dict(codon_usage, orient='index', columns=['Relative Frequency'])\n",
    "\n",
    "genome_sequences = load_genome(file_path)\n",
    "codon_usage_df = calculate_codon_usage(genome_sequences)\n",
    "print(codon_usage_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e7321686",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original Gene Sequence:   ATGGTTCCTCAGGCTCTCCTGTTTGTACCCCTTCTGGTTTTTCCATTGTGTTTTGGGAAATTCCCTATTTACACGATACCAGACAAGCTTGGTCCCTGGAGCCCGATTGACATACATCACCTCAGCTGCCCAAACAATTTGGTAGTGGAGGACGAAGGATGCACCAACCTGTCAGGGTTCTCCTACATGGAACTTAAAGTTGGATACATCTCAGCCATAAAAATGAACGGGTTCACTTGCACAGGCGTTGTGACGGAGGCTGAAACCTACACTAACTTCGTTGGTTATGTCACAACCACGTTCAAAAGAAAGCATTTCCGCCCAACACCAGATGCATGTAGAGCCGCGTACAACTGGAAGATGGCCGGTGACCCCAGATATGAAGAGTCTCTACACAATCCGTACCCTGACTACCACTGGCTTCGAACTGTAAAAACCACCAAGGAGTCTCTCGTTATCATATCTCCAAGTGTGGCAGATTTGGACCCATATGACAGATCCCTTCACTCGAGGGTCTTCCCTGGCGGGAATTGCTCAGGAGTAGCGGTGTCTTCTACCTACTGCTCCACTAACCACGATTACACCATTTGGATGCCCGAGAATCCGAGACTAGGGATGTCTTGTGACATTTTTACCAATAGTAGAGGGAAGAGAGCATCCAAAGGGAGTGAGACTTGCGGCTTTGTAGATGAAAGAGGCCTATATAAGTCTTTAAAAGGAGCATGCAAACTCAAGTTATGTGGAGTTCTAGGACTTAGACTTATGGATGGAACATGGGTCGCGATGCAAACATCAAATGAAACCAAATGGTGCCCTCCCGGTCAGTTGGTGAATTTGCACGACTTTCGCTCAGACGAAATTGAGCACCTTGTTGTAGAGGAGTTGGTCAAGAAGAGAGAGGAGTGTCTGGATGCACTAGAGTCCATCATGACCACCAAGTCAGTGAGTTTCAGACGTCTCAGTCATTTAAGAAAACTTGTCCCTGGGTTTGGAAAAGCATATACCATATTCAACAAGACCTTGATGGAAGCCGATGCTCACTACAAGTCAGTCAGAACTTGGAATGAGATCATCCCTTCAAAAGGGTGTTTAAGAGTTGGGGGGAGGTGTCATCCTCATGTAAACGGGGTATTTTTCAATGGTATAATATTAGGACCTGACGGCAATGTCTTAATCCCAGAGATGCAATCATCCCTCCTCCAGCAACATATGGAGTTGTTGGTATCCTCGGTTATCCCCCTTATGCACCCCCTGGCAGACCCGTCTACCGTTTTCAAGAACGGTGACGAGGCTGAGGATTTTGTTGAAGTTCACCTTCCCGATGTGCACGAACGGATCTCAGGAGTTGACTTGGGTCTCCCGAACTGGGGGAAGTATGTATTACTGAGTGCAGGGGCCCTGACTGCCTTGATGTTGATAATTTTCCTGATGACATGCTGGAGAAGAGTCAATCGATCGGAACCTACACAACACAATCTCAGAGGGACAGGGAGGGAGGTGTCAGTCACTCCCCAAAGCGGGAAGATCATATCTTCATGGGAATCATACAAGAGCGGGGGTGAGACCGGACTGTGA\n",
      "Optimized Gene Sequence:   ATGGTACCACAAGCATTATTATTTGTACCATTATTAGTATTTCCATTATGTTTTGGAAAATTTCCAATTTATACAATTCCAGATAAATTAGGACCATGGTCTCCAATTGATATTCATCATTTATCTTGTCCAAATAATTTAGTAGTAGAAGATGAAGGATGTACAAATTTATCTGGATTTTCTTATATGGAATTAAAAGTAGGATATATTTCTGCAATTAAAATGAATGGATTTACATGTACAGGAGTAGTAACAGAAGCAGAAACATATACAAATTTTGTAGGATATGTAACAACAACATTTAAACGTAAACATTTTCGTCCAACACCAGATGCATGTCGTGCAGCATATAATTGGAAAATGGCAGGAGATCCACGTTATGAAGAATCTTTACATAATCCATATCCAGATTATCATTGGTTACGTACAGTAAAAACAACAAAAGAATCTTTAGTAATTATTTCTCCATCTGTAGCAGATTTAGATCCATATGATCGTTCTTTACATTCTCGTGTATTTCCAGGAGGAAATTGTTCTGGAGTAGCAGTATCTTCTACATATTGTTCTACAAATCATGATTATACAATTTGGATGCCAGAAAATCCACGTTTAGGAATGTCTTGTGATATTTTTACAAATTCTCGTGGAAAACGTGCATCTAAAGGATCTGAAACATGTGGATTTGTAGATGAACGTGGATTATATAAATCTTTAAAAGGAGCATGTAAATTAAAATTATGTGGAGTATTAGGATTACGTTTAATGGATGGAACATGGGTAGCAATGCAAACATCTAATGAAACAAAATGGTGTCCACCAGGACAATTAGTAAATTTACATGATTTTCGTTCTGATGAAATTGAACATTTAGTAGTAGAAGAATTAGTAAAAAAACGTGAAGAATGTTTAGATGCATTAGAATCTATTATGACAACAAAATCTGTATCTTTTCGTCGTTTATCTCATTTACGTAAATTAGTACCAGGATTTGGAAAAGCATATACAATTTTTAATAAAACATTAATGGAAGCAGATGCACATTATAAATCTGTACGTACATGGAATGAAATTATTCCATCTAAAGGATGTTTACGTGTAGGAGGACGTTGTCATCCACATGTAAATGGAGTATTTTTTAATGGAATTATTTTAGGACCAGATGGAAATGTATTAATTCCAGAAATGCAATCTTCTTTATTACAACAACATATGGAATTATTAGTATCTTCTGTAATTCCATTAATGCATCCATTAGCAGATCCATCTACAGTATTTAAAAATGGAGATGAAGCAGAAGATTTTGTAGAAGTACATTTACCAGATGTACATGAACGTATTTCTGGAGTAGATTTAGGATTACCAAATTGGGGAAAATATGTATTATTATCTGCAGGAGCATTAACAGCATTAATGTTAATTATTTTTTTAATGACATGTTGGCGTCGTGTAAATCGTTCTGAACCAACACAACATAATTTACGTGGAACAGGACGTGAAGTATCTGTAACACCACAATCTGGAAAAATTATTTCTTCTTGGGAATCTTATAAATCTGGAGGAGAAACAGGATTA\n",
      "Highlighted Changes:      ATG[GTA][CCA][CAA][GCA][TTA][TTA]TTTGTA[CCA][TTA][TTA][GTA]TTTCCA[TTA]TGTTTT[GGA]AAA[TTT][CCA]ATT[TAT][ACA][ATT]CCA[GAT][AAA][TTA][GGA][CCA]TGG[TCT][CCA]ATT[GAT][ATT]CAT[CAT][TTA][TCT][TGT]CCA[AAT]AAT[TTA]GTA[GTA][GAA][GAT]GAAGGA[TGT][ACA][AAT][TTA][TCT][GGA][TTT][TCT][TAT]ATGGAA[TTA]AAA[GTA]GGA[TAT][ATT][TCT][GCA][ATT]AAAATG[AAT][GGA][TTT][ACA][TGT]ACA[GGA][GTA][GTA][ACA][GAA][GCA]GAA[ACA][TAT][ACA][AAT][TTT][GTA][GGA]TAT[GTA]ACA[ACA][ACA][TTT]AAA[CGT][AAA]CAT[TTT][CGT]CCAACACCAGATGCATGT[CGT][GCA][GCA][TAT][AAT]TGG[AAA]ATG[GCA][GGA][GAT][CCA][CGT]TATGAA[GAA]TCT[TTA][CAT]AAT[CCA][TAT][CCA][GAT][TAT][CAT]TGG[TTA][CGT][ACA]GTAAAA[ACA][ACA][AAA][GAA]TCT[TTA][GTA][ATT][ATT]TCTCCA[TCT][GTA]GCAGAT[TTA][GAT]CCATAT[GAT][CGT][TCT][TTA][CAT][TCT][CGT][GTA][TTT][CCA][GGA][GGA]AAT[TGT][TCT]GGAGTA[GCA][GTA]TCTTCT[ACA][TAT][TGT][TCT][ACA][AAT][CAT]GAT[TAT][ACA]ATTTGGATG[CCA][GAA]AAT[CCA][CGT][TTA][GGA]ATGTCTTGT[GAT]ATTTTT[ACA]AAT[TCT][CGT][GGA][AAA][CGT]GCA[TCT]AAA[GGA][TCT][GAA][ACA][TGT][GGA]TTTGTAGATGAA[CGT][GGA][TTA]TAT[AAA]TCTTTAAAAGGAGCA[TGT]AAA[TTA][AAA]TTATGTGGA[GTA][TTA]GGA[TTA][CGT][TTA]ATGGATGGAACATGG[GTA][GCA]ATGCAAACA[TCT]AATGAA[ACA]AAATGG[TGT][CCA][CCA][GGA][CAA][TTA][GTA]AAT[TTA][CAT][GAT]TTT[CGT][TCT][GAT]GAAATT[GAA][CAT][TTA][GTA]GTA[GAA][GAA][TTA][GTA][AAA][AAA][CGT][GAA][GAA]TGT[TTA]GATGCA[TTA][GAA][TCT][ATT]ATG[ACA][ACA][AAA][TCT][GTA][TCT][TTT][CGT]CGT[TTA][TCT]CATTTA[CGT]AAA[TTA][GTA][CCA][GGA]TTTGGAAAAGCATAT[ACA][ATT][TTT][AAT][AAA][ACA][TTA]ATGGAA[GCA]GAT[GCA][CAT][TAT][AAA][TCT][GTA][CGT][ACA]TGGAAT[GAA][ATT][ATT][CCA][TCT]AAA[GGA]TGTTTA[CGT][GTA][GGA][GGA][CGT]TGTCAT[CCA]CATGTA[AAT][GGA]GTATTT[TTT]AAT[GGA][ATT][ATT]TTAGGA[CCA][GAT][GGA]AAT[GTA]TTA[ATT]CCA[GAA]ATGCAA[TCT][TCT][TTA][TTA][CAA]CAACATATG[GAA][TTA][TTA]GTA[TCT][TCT][GTA][ATT][CCA][TTA]ATG[CAT][CCA][TTA]GCA[GAT][CCA]TCT[ACA][GTA][TTT][AAA][AAT][GGA][GAT][GAA][GCA][GAA]GATTTT[GTA]GAA[GTA][CAT][TTA][CCA]GAT[GTA][CAT]GAA[CGT][ATT][TCT]GGA[GTA][GAT][TTA][GGA][TTA][CCA][AAT]TGG[GGA][AAA]TATGTATTA[TTA][TCT]GCA[GGA][GCA][TTA][ACA][GCA][TTA]ATG[TTA][ATT]ATT[TTT][TTA]ATGACA[TGT]TGG[CGT][CGT][GTA]AAT[CGT][TCT]GAA[CCA]ACACAA[CAT]AAT[TTA][CGT][GGA]ACA[GGA][CGT][GAA][GTA][TCT][GTA][ACA][CCA]CAA[TCT][GGA][AAA][ATT][ATT]TCT[TCT]TGGGAA[TCT][TAT][AAA][TCT][GGA][GGA][GAA][ACA]GGA[TTA][]\n",
      "Number of changed codons: 354\n",
      "Number of same codons: 171\n"
     ]
    }
   ],
   "source": [
    "def design_optimized_gene(target_protein_seq, codon_usage_df):\n",
    "    codon_table = {\n",
    "        'A': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "        'C': ['TGT', 'TGC'],\n",
    "        'D': ['GAT', 'GAC'],\n",
    "        'E': ['GAA', 'GAG'],\n",
    "        'F': ['TTT', 'TTC'],\n",
    "        'G': ['GGT', 'GGC', 'GGA', 'GGG'],\n",
    "        'H': ['CAT', 'CAC'],\n",
    "        'I': ['ATT', 'ATC', 'ATA'],\n",
    "        'K': ['AAA', 'AAG'],\n",
    "        'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
    "        'M': ['ATG'],\n",
    "        'N': ['AAT', 'AAC'],\n",
    "        'P': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
    "        'Q': ['CAA', 'CAG'],\n",
    "        'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
    "        'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
    "        'T': ['ACT', 'ACC', 'ACA', 'ACG'],\n",
    "        'V': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "        'W': ['TGG'],\n",
    "        'Y': ['TAT', 'TAC'],\n",
    "        '*': ['TAA', 'TAG', 'TGA']\n",
    "    }\n",
    "    \n",
    "    optimized_gene = \"\"\n",
    "    for aa in target_protein_seq:\n",
    "        possible_codons = codon_table[aa]\n",
    "        best_codon = max(possible_codons, key=lambda codon: codon_usage_df.loc[codon, 'Relative Frequency'] if codon in codon_usage_df.index else 0)\n",
    "        optimized_gene += best_codon\n",
    "    \n",
    "    return optimized_gene\n",
    "\n",
    "target_protein_seq = \"MVPQALLFVPLLVFPLCFGKFPIYTIPDKLGPWSPIDIHHLSCPNNLVVEDEGCTNLSGFSYMELKVGYISAIKMNGFTCTGVVTEAETYTNFVGYVTTTFKRKHFRPTPDACRAAYNWKMAGDPRYEESLHNPYPDYHWLRTVKTTKESLVIISPSVADLDPYDRSLHSRVFPGGNCSGVAVSSTYCSTNHDYTIWMPENPRLGMSCDIFTNSRGKRASKGSETCGFVDERGLYKSLKGACKLKLCGVLGLRLMDGTWVAMQTSNETKWCPPGQLVNLHDFRSDEIEHLVVEELVKKREECLDALESIMTTKSVSFRRLSHLRKLVPGFGKAYTIFNKTLMEADAHYKSVRTWNEIIPSKGCLRVGGRCHPHVNGVFFNGIILGPDGNVLIPEMQSSLLQQHMELLVSSVIPLMHPLADPSTVFKNGDEAEDFVEVHLPDVHERISGVDLGLPNWGKYVLLSAGALTALMLIIFLMTCWRRVNRSEPTQHNLRGTGREVSVTPQSGKIISSWESYKSGGETGL\"\n",
    "\n",
    "original_gene_seq = \"ATGGTTCCTCAGGCTCTCCTGTTTGTACCCCTTCTGGTTTTTCCATTGTGTTTTGGGAAATTCCCTATTTACACGATACCAGACAAGCTTGGTCCCTGGAGCCCGATTGACATACATCACCTCAGCTGCCCAAACAATTTGGTAGTGGAGGACGAAGGATGCACCAACCTGTCAGGGTTCTCCTACATGGAACTTAAAGTTGGATACATCTCAGCCATAAAAATGAACGGGTTCACTTGCACAGGCGTTGTGACGGAGGCTGAAACCTACACTAACTTCGTTGGTTATGTCACAACCACGTTCAAAAGAAAGCATTTCCGCCCAACACCAGATGCATGTAGAGCCGCGTACAACTGGAAGATGGCCGGTGACCCCAGATATGAAGAGTCTCTACACAATCCGTACCCTGACTACCACTGGCTTCGAACTGTAAAAACCACCAAGGAGTCTCTCGTTATCATATCTCCAAGTGTGGCAGATTTGGACCCATATGACAGATCCCTTCACTCGAGGGTCTTCCCTGGCGGGAATTGCTCAGGAGTAGCGGTGTCTTCTACCTACTGCTCCACTAACCACGATTACACCATTTGGATGCCCGAGAATCCGAGACTAGGGATGTCTTGTGACATTTTTACCAATAGTAGAGGGAAGAGAGCATCCAAAGGGAGTGAGACTTGCGGCTTTGTAGATGAAAGAGGCCTATATAAGTCTTTAAAAGGAGCATGCAAACTCAAGTTATGTGGAGTTCTAGGACTTAGACTTATGGATGGAACATGGGTCGCGATGCAAACATCAAATGAAACCAAATGGTGCCCTCCCGGTCAGTTGGTGAATTTGCACGACTTTCGCTCAGACGAAATTGAGCACCTTGTTGTAGAGGAGTTGGTCAAGAAGAGAGAGGAGTGTCTGGATGCACTAGAGTCCATCATGACCACCAAGTCAGTGAGTTTCAGACGTCTCAGTCATTTAAGAAAACTTGTCCCTGGGTTTGGAAAAGCATATACCATATTCAACAAGACCTTGATGGAAGCCGATGCTCACTACAAGTCAGTCAGAACTTGGAATGAGATCATCCCTTCAAAAGGGTGTTTAAGAGTTGGGGGGAGGTGTCATCCTCATGTAAACGGGGTATTTTTCAATGGTATAATATTAGGACCTGACGGCAATGTCTTAATCCCAGAGATGCAATCATCCCTCCTCCAGCAACATATGGAGTTGTTGGTATCCTCGGTTATCCCCCTTATGCACCCCCTGGCAGACCCGTCTACCGTTTTCAAGAACGGTGACGAGGCTGAGGATTTTGTTGAAGTTCACCTTCCCGATGTGCACGAACGGATCTCAGGAGTTGACTTGGGTCTCCCGAACTGGGGGAAGTATGTATTACTGAGTGCAGGGGCCCTGACTGCCTTGATGTTGATAATTTTCCTGATGACATGCTGGAGAAGAGTCAATCGATCGGAACCTACACAACACAATCTCAGAGGGACAGGGAGGGAGGTGTCAGTCACTCCCCAAAGCGGGAAGATCATATCTTCATGGGAATCATACAAGAGCGGGGGTGAGACCGGACTGTGA\"\n",
    "\n",
    "optimized_gene_seq = design_optimized_gene(target_protein_seq, codon_usage_df)\n",
    "\n",
    "highlighted_changes = \"\"\n",
    "num_changed_codons = 0\n",
    "num_same_codons = 0\n",
    "\n",
    "for i in range(0, len(original_gene_seq), 3):\n",
    "    original_codon = original_gene_seq[i:i+3]\n",
    "    optimized_codon = optimized_gene_seq[i:i+3]\n",
    "    if original_codon != optimized_codon:\n",
    "        highlighted_changes += f\"[{optimized_codon}]\"\n",
    "        num_changed_codons += 1\n",
    "    else:\n",
    "        highlighted_changes += optimized_codon\n",
    "        num_same_codons += 1\n",
    "\n",
    "print(\"Original Gene Sequence:  \", original_gene_seq)\n",
    "print(\"Optimized Gene Sequence:  \", optimized_gene_seq)\n",
    "print(\"Highlighted Changes:     \", highlighted_changes)\n",
    "print(f\"Number of changed codons: {num_changed_codons}\")\n",
    "print(f\"Number of same codons: {num_same_codons}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d23668bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Oryginalny stosunek par CG: 0.48\n",
      "Zoptymalizowany stosunek par CG: 0.31\n"
     ]
    }
   ],
   "source": [
    "def cg_ratio(sequence):\n",
    "    return (sequence.count('C') + sequence.count('G')) / len(sequence)\n",
    "\n",
    "original_cg_ratio = cg_ratio(original_gene_seq)\n",
    "optimized_cg_ratio = cg_ratio(optimized_gene_seq)\n",
    "\n",
    "print(f\"Oryginalny stosunek par CG: {original_cg_ratio:.2f}\")\n",
    "print(f\"Zoptymalizowany stosunek par CG: {optimized_cg_ratio:.2f}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
